src/PrestaShopBundle/Form/Admin/Type/ShopChoiceTreeType.php

```diff
@@ -38,25 +40,33 @@
 class ShopChoiceTreeType extends AbstractType
 {
     /**
-     * @var array
+     * @var FormChoiceProviderInterface
      */
-    private $shopTreeChoices;
+    private $shopTreeChoiceProvider;
 
     /**
      * @var DataTransformerInterface
      */
     private $stringArrayToIntegerArrayDataTransformer;
 
     /**
-     * @param array $shopTreeChoices
+     * @var ShopContextInterface
+     */
+    private $shopContext;
+
+    /**
+     * @param FormChoiceProviderInterface $shopTreeChoiceProvider
      * @param DataTransformerInterface $stringArrayToIntegerArrayDataTransformer
+     * @param ShopContextInterface $shopContext
      */
     public function __construct(
-        array $shopTreeChoices,
-        DataTransformerInterface $stringArrayToIntegerArrayDataTransformer
+        FormChoiceProviderInterface $shopTreeChoiceProvider,
```

Comment:
```suggestion
        ShopTreeChoiceProvider $shopTreeChoiceProvider,
```
If you use the implementation class instead then you can rely on autowiring, I think it's a nice way to handle this in form types, it doesn't respect the Dependency inversion principle but it's quite useful for simple configuration. I wonder if we should accept this compromise 🤔

Reply:
yeah I considered it, but the implementation is in Adapter ns